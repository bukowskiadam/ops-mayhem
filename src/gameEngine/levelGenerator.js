const LevelGenerator = (
  {
    timeOfRest,
    timeToFix,
    boardSize,
    maxOverdue,
    pointsMultiplier,
  },
  {
    timeDecrease,
    sizeIncrease,
    pointsIncrease,
  },
) => {
  const make = (level) => {
    const timeDecreaseFactor = Math.pow(timeDecrease, level - 1);
    const sizeIncreaseFactor = Math.pow(sizeIncrease, level - 1);

    return ({
      timeOfRest: timeOfRest * timeDecreaseFactor,
      timeToFix: timeToFix * timeDecreaseFactor,
      boardSize: Math.floor(boardSize * sizeIncreaseFactor),
      maxOverdue: Math.floor(maxOverdue * sizeIncreaseFactor),
      pointsMultiplier: pointsMultiplier * Math.pow(pointsIncrease, level - 1),
    });
  };

  return {
    make,
  };
};

LevelGenerator.PREDEFINED = {
  noob: () => LevelGenerator(
    {
      timeOfRest: 750,
      timeToFix: 1500,
      boardSize: 3,
      maxOverdue: 3,
      pointsMultiplier: 1,
    },
    {
      timeDecrease: 1 / 1.1,
      sizeIncrease: 1.05,
      pointsIncrease: 1.1,
    },
  ),
  pro: () => LevelGenerator(
    {
      timeOfRest: 500,
      timeToFix: 1250,
      boardSize: 4,
      maxOverdue: 4,
      pointsMultiplier: 1.2,
    },
    {
      timeDecrease: 1 / 1.2,
      sizeIncrease: 1.05,
      pointsIncrease: 1.2,
    },
  ),
  filip: () => LevelGenerator(
    {
      timeOfRest: 300,
      timeToFix: 1000,
      boardSize: 5,
      maxOverdue: 3,
      pointsMultiplier: 3,
    },
    {
      timeDecrease: 1 / 1.5,
      sizeIncrease: 1.1,
      pointsIncrease: 2,
    },
  ),
};
export { LevelGenerator };
