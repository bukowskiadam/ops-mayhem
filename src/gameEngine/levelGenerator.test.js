import { LevelGenerator } from './levelGenerator';

describe('Level generator', () => {
  const makeParamsSnapshotTest = (generator, levelsCount) => {
    const levels = [];

    for (let i = 1; i < levelsCount; i += 1) {
      levels.push(generator.make(i));
    }

    expect(levels).toMatchSnapshot();
  };

  const calculateMaxPointsForLevel = (generator, levelsCount) => {
    const points = [];
    let total = 0;

    for (let i = 1; i < levelsCount; i += 1) {
      const { pointsMultiplier, timeToFix, boardSize } = generator.make(i);
      const fields = boardSize * boardSize;
      const levelPoints = timeToFix * pointsMultiplier * fields;
      total += levelPoints;

      points.push({ levelPoints, total });
    }

    expect(points).toMatchSnapshot();
  };

  it('creates levels params for a noob gameplay', () => {
    const generator = LevelGenerator.PREDEFINED.noob();

    makeParamsSnapshotTest(generator, 20);
  });

  it('creates levels params for a pro gameplay', () => {
    const generator = LevelGenerator.PREDEFINED.pro();

    makeParamsSnapshotTest(generator, 20);
  });

  it('creates levels params for a filip gameplay', () => {
    const generator = LevelGenerator.PREDEFINED.filip();

    makeParamsSnapshotTest(generator, 20);
  });

  it('calculate points for a noob gameplay', () => {
    const generator = LevelGenerator.PREDEFINED.noob();

    calculateMaxPointsForLevel(generator, 20);
  });

  it('calculate points for a pro gameplay', () => {
    const generator = LevelGenerator.PREDEFINED.pro();

    calculateMaxPointsForLevel(generator, 20);
  });

  it('calculate points for a filip gameplay', () => {
    const generator = LevelGenerator.PREDEFINED.filip();

    calculateMaxPointsForLevel(generator, 20);
  });
});
