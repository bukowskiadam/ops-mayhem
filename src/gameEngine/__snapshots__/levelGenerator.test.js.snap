// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Level generator calculate points for a filip gameplay 1`] = `
Array [
  Object {
    "levelPoints": 75000,
    "total": 75000,
  },
  Object {
    "levelPoints": 100000,
    "total": 175000,
  },
  Object {
    "levelPoints": 192000,
    "total": 367000,
  },
  Object {
    "levelPoints": 255999.99999999994,
    "total": 623000,
  },
  Object {
    "levelPoints": 464592.5925925925,
    "total": 1087592.5925925926,
  },
  Object {
    "levelPoints": 809086.4197530863,
    "total": 1896679.0123456789,
  },
  Object {
    "levelPoints": 1078781.893004115,
    "total": 2975460.905349794,
  },
  Object {
    "levelPoints": 1820444.4444444436,
    "total": 4795905.349794237,
  },
  Object {
    "levelPoints": 2996616.3694558744,
    "total": 7792521.719250111,
  },
  Object {
    "levelPoints": 4834541.076055476,
    "total": 12627062.795305587,
  },
  Object {
    "levelPoints": 7671337.905807038,
    "total": 20298400.701112624,
  },
  Object {
    "levelPoints": 13922057.680909066,
    "total": 34220458.38202169,
  },
  Object {
    "levelPoints": 21309271.9605751,
    "total": 55529730.342596784,
  },
  Object {
    "levelPoints": 36494101.313221954,
    "total": 92023831.65581873,
  },
  Object {
    "levelPoints": 54551736.21907225,
    "total": 146575567.87489098,
  },
  Object {
    "levelPoints": 89797096.65690906,
    "total": 236372664.53180003,
  },
  Object {
    "levelPoints": 144872649.2731466,
    "total": 381245313.80494666,
  },
  Object {
    "levelPoints": 249436379.60252512,
    "total": 630681693.4074718,
  },
  Object {
    "levelPoints": 387923457.557847,
    "total": 1018605150.9653188,
  },
]
`;

exports[`Level generator calculate points for a noob gameplay 1`] = `
Array [
  Object {
    "levelPoints": 13500,
    "total": 13500,
  },
  Object {
    "levelPoints": 13500,
    "total": 27000,
  },
  Object {
    "levelPoints": 13500,
    "total": 40500,
  },
  Object {
    "levelPoints": 13500.000000000002,
    "total": 54000,
  },
  Object {
    "levelPoints": 13500.000000000005,
    "total": 67500,
  },
  Object {
    "levelPoints": 13500,
    "total": 81000,
  },
  Object {
    "levelPoints": 24000.000000000007,
    "total": 105000,
  },
  Object {
    "levelPoints": 24000.00000000001,
    "total": 129000.00000000001,
  },
  Object {
    "levelPoints": 24000.00000000001,
    "total": 153000.00000000003,
  },
  Object {
    "levelPoints": 24000.00000000001,
    "total": 177000.00000000003,
  },
  Object {
    "levelPoints": 24000.000000000015,
    "total": 201000.00000000006,
  },
  Object {
    "levelPoints": 37500.000000000015,
    "total": 238500.00000000006,
  },
  Object {
    "levelPoints": 37500.00000000002,
    "total": 276000.00000000006,
  },
  Object {
    "levelPoints": 37500.00000000003,
    "total": 313500.0000000001,
  },
  Object {
    "levelPoints": 37500.00000000002,
    "total": 351000.0000000001,
  },
  Object {
    "levelPoints": 54000.000000000044,
    "total": 405000.0000000002,
  },
  Object {
    "levelPoints": 54000.00000000003,
    "total": 459000.00000000023,
  },
  Object {
    "levelPoints": 54000.000000000044,
    "total": 513000.0000000003,
  },
  Object {
    "levelPoints": 73500.00000000007,
    "total": 586500.0000000003,
  },
]
`;

exports[`Level generator calculate points for a pro gameplay 1`] = `
Array [
  Object {
    "levelPoints": 24000,
    "total": 24000,
  },
  Object {
    "levelPoints": 24000,
    "total": 48000,
  },
  Object {
    "levelPoints": 24000.000000000004,
    "total": 72000,
  },
  Object {
    "levelPoints": 23999.999999999996,
    "total": 96000,
  },
  Object {
    "levelPoints": 24000.000000000004,
    "total": 120000,
  },
  Object {
    "levelPoints": 37500,
    "total": 157500,
  },
  Object {
    "levelPoints": 37500,
    "total": 195000,
  },
  Object {
    "levelPoints": 37499.99999999999,
    "total": 232500,
  },
  Object {
    "levelPoints": 37500,
    "total": 270000,
  },
  Object {
    "levelPoints": 54000.00000000001,
    "total": 324000,
  },
  Object {
    "levelPoints": 54000,
    "total": 378000,
  },
  Object {
    "levelPoints": 54000,
    "total": 432000,
  },
  Object {
    "levelPoints": 73500,
    "total": 505500,
  },
  Object {
    "levelPoints": 73499.99999999999,
    "total": 579000,
  },
  Object {
    "levelPoints": 73500,
    "total": 652500,
  },
  Object {
    "levelPoints": 96000.00000000001,
    "total": 748500,
  },
  Object {
    "levelPoints": 96000.00000000001,
    "total": 844500,
  },
  Object {
    "levelPoints": 121499.99999999999,
    "total": 966000,
  },
  Object {
    "levelPoints": 121500.00000000001,
    "total": 1087500,
  },
]
`;

exports[`Level generator creates levels params for a filip gameplay 1`] = `
Array [
  Object {
    "boardSize": 5,
    "maxOverdue": 3,
    "pointsMultiplier": 3,
    "timeOfRest": 300,
    "timeToFix": 1000,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 3,
    "pointsMultiplier": 6,
    "timeOfRest": 200,
    "timeToFix": 666.6666666666666,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 3,
    "pointsMultiplier": 12,
    "timeOfRest": 133.33333333333331,
    "timeToFix": 444.4444444444444,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 3,
    "pointsMultiplier": 24,
    "timeOfRest": 88.88888888888887,
    "timeToFix": 296.29629629629625,
  },
  Object {
    "boardSize": 7,
    "maxOverdue": 4,
    "pointsMultiplier": 48,
    "timeOfRest": 59.259259259259245,
    "timeToFix": 197.53086419753083,
  },
  Object {
    "boardSize": 8,
    "maxOverdue": 4,
    "pointsMultiplier": 96,
    "timeOfRest": 39.506172839506156,
    "timeToFix": 131.68724279835388,
  },
  Object {
    "boardSize": 8,
    "maxOverdue": 5,
    "pointsMultiplier": 192,
    "timeOfRest": 26.337448559670776,
    "timeToFix": 87.79149519890258,
  },
  Object {
    "boardSize": 9,
    "maxOverdue": 5,
    "pointsMultiplier": 384,
    "timeOfRest": 17.558299039780515,
    "timeToFix": 58.52766346593505,
  },
  Object {
    "boardSize": 10,
    "maxOverdue": 6,
    "pointsMultiplier": 768,
    "timeOfRest": 11.705532693187008,
    "timeToFix": 39.018442310623364,
  },
  Object {
    "boardSize": 11,
    "maxOverdue": 7,
    "pointsMultiplier": 1536,
    "timeOfRest": 7.803688462124673,
    "timeToFix": 26.012294873748907,
  },
  Object {
    "boardSize": 12,
    "maxOverdue": 7,
    "pointsMultiplier": 3072,
    "timeOfRest": 5.202458974749781,
    "timeToFix": 17.341529915832606,
  },
  Object {
    "boardSize": 14,
    "maxOverdue": 8,
    "pointsMultiplier": 6144,
    "timeOfRest": 3.4683059831665206,
    "timeToFix": 11.561019943888402,
  },
  Object {
    "boardSize": 15,
    "maxOverdue": 9,
    "pointsMultiplier": 12288,
    "timeOfRest": 2.3122039887776804,
    "timeToFix": 7.707346629258934,
  },
  Object {
    "boardSize": 17,
    "maxOverdue": 10,
    "pointsMultiplier": 24576,
    "timeOfRest": 1.5414693258517869,
    "timeToFix": 5.138231086172623,
  },
  Object {
    "boardSize": 18,
    "maxOverdue": 11,
    "pointsMultiplier": 49152,
    "timeOfRest": 1.0276462172345244,
    "timeToFix": 3.4254873907817482,
  },
  Object {
    "boardSize": 20,
    "maxOverdue": 12,
    "pointsMultiplier": 98304,
    "timeOfRest": 0.6850974781563496,
    "timeToFix": 2.2836582605211655,
  },
  Object {
    "boardSize": 22,
    "maxOverdue": 13,
    "pointsMultiplier": 196608,
    "timeOfRest": 0.456731652104233,
    "timeToFix": 1.5224388403474434,
  },
  Object {
    "boardSize": 25,
    "maxOverdue": 15,
    "pointsMultiplier": 393216,
    "timeOfRest": 0.3044877680694887,
    "timeToFix": 1.0149592268982957,
  },
  Object {
    "boardSize": 27,
    "maxOverdue": 16,
    "pointsMultiplier": 786432,
    "timeOfRest": 0.2029918453796591,
    "timeToFix": 0.6766394845988637,
  },
]
`;

exports[`Level generator creates levels params for a noob gameplay 1`] = `
Array [
  Object {
    "boardSize": 3,
    "maxOverdue": 3,
    "pointsMultiplier": 1,
    "timeOfRest": 750,
    "timeToFix": 1500,
  },
  Object {
    "boardSize": 3,
    "maxOverdue": 3,
    "pointsMultiplier": 1.1,
    "timeOfRest": 681.8181818181818,
    "timeToFix": 1363.6363636363635,
  },
  Object {
    "boardSize": 3,
    "maxOverdue": 3,
    "pointsMultiplier": 1.2100000000000002,
    "timeOfRest": 619.8347107438016,
    "timeToFix": 1239.6694214876031,
  },
  Object {
    "boardSize": 3,
    "maxOverdue": 3,
    "pointsMultiplier": 1.3310000000000004,
    "timeOfRest": 563.4861006761832,
    "timeToFix": 1126.9722013523665,
  },
  Object {
    "boardSize": 3,
    "maxOverdue": 3,
    "pointsMultiplier": 1.4641000000000006,
    "timeOfRest": 512.260091523803,
    "timeToFix": 1024.520183047606,
  },
  Object {
    "boardSize": 3,
    "maxOverdue": 3,
    "pointsMultiplier": 1.6105100000000006,
    "timeOfRest": 465.69099229436625,
    "timeToFix": 931.3819845887325,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 1.7715610000000008,
    "timeOfRest": 423.355447540333,
    "timeToFix": 846.710895080666,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 1.9487171000000012,
    "timeOfRest": 384.86858867303,
    "timeToFix": 769.73717734606,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 2.1435888100000016,
    "timeOfRest": 349.88053515729996,
    "timeToFix": 699.7610703145999,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 2.357947691000002,
    "timeOfRest": 318.0732137793636,
    "timeToFix": 636.1464275587272,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 2.5937424601000023,
    "timeOfRest": 289.15746707214873,
    "timeToFix": 578.3149341442975,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 2.8531167061100025,
    "timeOfRest": 262.87042461104426,
    "timeToFix": 525.7408492220885,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 3.138428376721003,
    "timeOfRest": 238.9731132827675,
    "timeToFix": 477.946226565535,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 3.452271214393104,
    "timeOfRest": 217.24828480251594,
    "timeToFix": 434.4965696050319,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 3.7974983358324144,
    "timeOfRest": 197.49844072955992,
    "timeToFix": 394.99688145911983,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 6,
    "pointsMultiplier": 4.177248169415656,
    "timeOfRest": 179.54403702687264,
    "timeToFix": 359.0880740537453,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 6,
    "pointsMultiplier": 4.594972986357222,
    "timeOfRest": 163.22185184261147,
    "timeToFix": 326.44370368522294,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 6,
    "pointsMultiplier": 5.054470284992945,
    "timeOfRest": 148.38350167510134,
    "timeToFix": 296.7670033502027,
  },
  Object {
    "boardSize": 7,
    "maxOverdue": 7,
    "pointsMultiplier": 5.55991731349224,
    "timeOfRest": 134.8940924319103,
    "timeToFix": 269.7881848638206,
  },
]
`;

exports[`Level generator creates levels params for a pro gameplay 1`] = `
Array [
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 1.2,
    "timeOfRest": 500,
    "timeToFix": 1250,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 1.44,
    "timeOfRest": 416.6666666666667,
    "timeToFix": 1041.6666666666667,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 1.728,
    "timeOfRest": 347.2222222222223,
    "timeToFix": 868.0555555555557,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 2.0735999999999994,
    "timeOfRest": 289.35185185185185,
    "timeToFix": 723.3796296296297,
  },
  Object {
    "boardSize": 4,
    "maxOverdue": 4,
    "pointsMultiplier": 2.48832,
    "timeOfRest": 241.1265432098766,
    "timeToFix": 602.8163580246915,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 2.9859839999999993,
    "timeOfRest": 200.9387860082305,
    "timeToFix": 502.34696502057625,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 3.583180799999999,
    "timeOfRest": 167.4489883401921,
    "timeToFix": 418.6224708504802,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 4.2998169599999985,
    "timeOfRest": 139.54082361682674,
    "timeToFix": 348.85205904206686,
  },
  Object {
    "boardSize": 5,
    "maxOverdue": 5,
    "pointsMultiplier": 5.159780351999998,
    "timeOfRest": 116.28401968068897,
    "timeToFix": 290.7100492017224,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 6,
    "pointsMultiplier": 6.191736422399998,
    "timeOfRest": 96.90334973390748,
    "timeToFix": 242.2583743347687,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 6,
    "pointsMultiplier": 7.430083706879996,
    "timeOfRest": 80.75279144492289,
    "timeToFix": 201.88197861230725,
  },
  Object {
    "boardSize": 6,
    "maxOverdue": 6,
    "pointsMultiplier": 8.916100448255996,
    "timeOfRest": 67.29399287076907,
    "timeToFix": 168.2349821769227,
  },
  Object {
    "boardSize": 7,
    "maxOverdue": 7,
    "pointsMultiplier": 10.699320537907195,
    "timeOfRest": 56.07832739230757,
    "timeToFix": 140.19581848076893,
  },
  Object {
    "boardSize": 7,
    "maxOverdue": 7,
    "pointsMultiplier": 12.839184645488633,
    "timeOfRest": 46.73193949358964,
    "timeToFix": 116.8298487339741,
  },
  Object {
    "boardSize": 7,
    "maxOverdue": 7,
    "pointsMultiplier": 15.40702157458636,
    "timeOfRest": 38.94328291132471,
    "timeToFix": 97.35820727831177,
  },
  Object {
    "boardSize": 8,
    "maxOverdue": 8,
    "pointsMultiplier": 18.48842588950363,
    "timeOfRest": 32.45273575943726,
    "timeToFix": 81.13183939859314,
  },
  Object {
    "boardSize": 8,
    "maxOverdue": 8,
    "pointsMultiplier": 22.186111067404358,
    "timeOfRest": 27.043946466197713,
    "timeToFix": 67.60986616549428,
  },
  Object {
    "boardSize": 9,
    "maxOverdue": 9,
    "pointsMultiplier": 26.623333280885223,
    "timeOfRest": 22.53662205516476,
    "timeToFix": 56.341555137911904,
  },
  Object {
    "boardSize": 9,
    "maxOverdue": 9,
    "pointsMultiplier": 31.94799993706227,
    "timeOfRest": 18.78051837930397,
    "timeToFix": 46.951295948259926,
  },
]
`;
